src/frontend/context/SecurityContext.tsx

```diff
@@ -0,0 +1,155 @@
+import * as React from "react";
+
+import { OBSCURE_KEY, OBSCURE_PASSCODE, PASSWORD_KEY } from "../constants";
+import createPersistedState from "../hooks/usePersistedState";
+
+type AuthState = "unauthenticated" | "authenticated" | "obscured";
+
+type AuthSetters =
+  | { type: "passcode"; value: string | null }
+  //This is to set up the future use of the obscure pass beings set by the user. if the `value` is undefined, the default obscure pass is being used
+  | { type: "obscure"; value?: string | null };
+
+type AuthValuesSet = {
+  passcodeSet: boolean;
+  obscureSet: boolean;
+};
+
+type SecurityContextType = {
+  authValuesSet: AuthValuesSet;
+  setAuthValues: (val: AuthSetters) => void;
+  authenticate: (val: string | null, validateOnly?: boolean) => boolean;
+  authState: AuthState;
+};
+
+const DefaultState: SecurityContextType = {
+  authValuesSet: { passcodeSet: false, obscureSet: false },
+  setAuthValues: () => {},
+  authenticate: () => false,
+  authState: "unauthenticated",
+};
+
+export const SecurityContext = React.createContext<SecurityContextType>(
+  DefaultState
+);
+
+const usePersistedPasscodeState = createPersistedState(PASSWORD_KEY);
+const usePersistedObscureState = createPersistedState(OBSCURE_KEY);
+
+export const SecurityProvider = ({
+  children,
+}: {
+  children: React.ReactNode;
+}) => {
+  const [authState, setAuthState] = React.useState<AuthState>(
+    "unauthenticated"
+  );
+
+  const [passcode, passCodestatus, setPasscode] = usePersistedPasscodeState<
+    string | null
+  >(null);
+
+  const [obscureCode, , setobscureCode] = usePersistedObscureState<
+    string | null
+  >(null);
+
+  // if passcode is unset, we want to makes sure that obscure code is unset, and that the user is 'authenticated' aka does not need a passcode to enter the app.
+  React.useEffect(() => {
+    if (passcode === null) {
+      setobscureCode(null);
+      setAuthState("authenticated");
+    }
+  }, [passcode]);
+
+  const setPasscodeWithValidation = React.useCallback(
+    (passcodeValue: string | null) => {
+      if (passcodeValue === OBSCURE_PASSCODE) {
+        throw new Error("passcode is reserved");
+      }
+
+      if (!validPasscode(passcodeValue)) {
+        throw new Error("passcode not valid");
+      }
+
+      setPasscode(passcodeValue);
+    },
+    [OBSCURE_PASSCODE]
+  );
+
+  const setObscureCodeWithValidation = React.useCallback(
+    (newObscureVal: string | null | undefined) => {
```

Comment:
when using this function, currently nowhere in the app passes `newObscureValue` because when the value is undefined it sets it to the default. There is no UI to set the obscureCode, but this architecture will make it easy to set obscureCode if we choose to do that.